### PR TITLE
Add generic neighbor transformations and use them to fix periodic coordinates for plex

### PR DESCRIPTION
**Pull request by tisaac**
There are four parts to this PR

# 1. Small compiler warning / indent fixes for p6est

Self-explanatory.

# 2. Add operations on pure coordinates

- `p4est_coordinates_compare()`
- `p4est_coordinates_transform_face()`
- `p8est_coordinates_transform_edge()`

From the commit messages:

> It is insufficient just to transform nodes because nodes are restricted to being at valid quadrant corners, so not all coordinates can be represented by nodes.
>
> Likewise, the coordinates can be sorted in Morton order even without a level attached, which is useful for canonicalizing points between trees, so we add this capability.

# 3. Add generic neighbor transformations

- typedef `p4est_neighbor_transform_t`
- `p4est_connectivity_get_neighbor_transforms()`
- `p4est_neighbor_transform_coordinates()` and `p4est_neighbor_transform_coordinates_reverse()`
- `p4est_neighbor_transform_quadrant()` and `p4est_neighbor_transform_quadrant_reverse()`

From the commit messages:

> Add a neighbor transform that generalizes all other inter-tree transformation objects.
>
> - This makes it much easier to write loops over all trees that touch a {face, edge, corner}
> - The data structure, while not as compressed as the others, is more interpretable: each transformation is clearly composed from shifts, permutations, and sign changes.
> - A test is added to make sure the generic neighbor transformation does the same thing as the individual transformations for every interface of every named connectivity.

# 4. Fix vertex coordinates calculations in `p4est_get_plex_data()`

From the commit messages:

> Guarantee that all cells agree about the coordinates of a vertex in plex data
>
> Use newly added neighbor transformations to canonicalize the (tree, qcoords) of each vertex and then apply qcoords_to_vertex to determine the coordinates.
> A periodic connectivity is added test_plex2.c to exercise this capability.

### CODE REVIEW

**Comment by cburstedde** [comment]
Thanks! I'll need into next week to look at this. I'm trusting that it remains legal and supported to have a periodic tree corner for which each tree has different vertex coordinates.

**Comment by tisaac** [comment]
> Thanks! I'll need into next week to look at this. I'm trusting that it remains legal and supported to have a periodic tree corner for which each tree has different vertex coordinates.

Yes, this doesn't change what is legal in a connectivity, only what is canonically decided to be the coordinates associated with points on the boundary between trees with differing coordinates, and only in the plex interface.

File: src/p6est.c [hunk 1]
```diff
@@ -162,8 +162,8 @@ p6est_connectivity_memory_used (p6est_connectivity_t * conn)
 {
   return
     p4est_connectivity_memory_used (conn->conn4) +
-    conn->top_vertices == NULL ? 0 :
-    (conn->conn4->num_vertices * 3 * sizeof (double));
+    ((conn->top_vertices == NULL) ? 0 :
+     (conn->conn4->num_vertices * 3 * sizeof (double)));
```
Comment (cburstedde) [inline src/p6est.c:166]: Good catch.

**Comment by cburstedde** [comment]
Sorry it took me a long time to get back. Would you think this change may degrade the runtime of p4est_quadrant_compare?

Other than that, I'm fine with everything.

**Comment by tisaac** [comment]
I took a look at the assembly that was generated before and after, and this is my understanding of the difference:

- If the coordinates are equal, the operations are essentially the same.
- If the coordinates are not equal, the new version has one additional `je` comparison of the outputs of `p4est_coordinates_compare()` to zero.

I have not compared the timings.

**Comment by cburstedde** [comment]
Thanks for looking into it, T.

**Pull request merged**